File: find_dups_cp/find_dups.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <filesystem>
#include <unordered_map>
#include <algorithm>
#include <chrono>
#include <iomanip>
#include <sstream>
#include <thread>
#include <future>
#include <mutex>
#include <queue>
#include <condition_variable>
#include <atomic>
#include <sys/stat.h>
#include <openssl/evp.h>

namespace fs = std::filesystem;

using Path = fs::path;
using Hash = std::string;
using TimePoint = std::chrono::system_clock::time_point;
using Duration = std::chrono::duration<double>;

struct FileInfo {
    uint64_t id;
    Path path;
    uint64_t size;
    TimePoint mod_time;
    TimePoint birth_time;
    Hash hash;
};

const size_t BUFFER_SIZE = 65536;
const unsigned int NUM_WORKERS = std::thread::hardware_concurrency();

std::string format_time(TimePoint tp) {
    auto tt = std::chrono::system_clock::to_time_t(tp);
    std::tm tm = *std::gmtime(&tt);
    std::ostringstream oss;
    oss << std::put_time(&tm, "%Y-%m-%dT%H:%M:%S") << "Z";
    return oss.str();
}

Hash compute_sha256(const Path& path) {
    std::ifstream file(path, std::ios::binary);
    if (!file.is_open()) return "";

    EVP_MD_CTX* ctx = EVP_MD_CTX_new();
    EVP_DigestInit_ex(ctx, EVP_sha256(), nullptr);

    char buffer[BUFFER_SIZE];
    while (file.read(buffer, sizeof(buffer))) {
        EVP_DigestUpdate(ctx, buffer, sizeof(buffer));
    }
    EVP_DigestUpdate(ctx, buffer, file.gcount());

    unsigned char hash[EVP_MAX_MD_SIZE];
    unsigned int hash_len = 0;
    EVP_DigestFinal_ex(ctx, hash, &hash_len);
    EVP_MD_CTX_free(ctx);

    std::ostringstream oss;
    for (unsigned int i = 0; i < hash_len; i++)
        oss << std::hex << std::setw(2) << std::setfill('0') << (int)hash[i];
    return oss.str();
}

std::vector<FileInfo> collect_files(const std::vector<Path>& roots, uint64_t& scanned) {
    std::vector<FileInfo> files;
    uint64_t next_id = 1;

    for (const auto& root : roots) {
        try {
            for (auto& entry : fs::recursive_directory_iterator(root,
                            fs::directory_options::skip_permission_denied)) {
                if (entry.is_symlink()) continue;
                if (!entry.is_regular_file()) continue;

                auto path = entry.path();
                auto size = entry.file_size();
                if (size == 0) continue; // skip zero-byte files

                auto mod_time = entry.last_write_time();
                auto sys_mtime = std::chrono::time_point_cast<std::chrono::system_clock::duration>(
                                    mod_time - fs::file_time_type::clock::now() +
                                    std::chrono::system_clock::now());

                struct stat stat_buf;
                if (stat(path.c_str(), &stat_buf) == 0) {
                    auto birth = std::chrono::system_clock::from_time_t(stat_buf.st_birthtime);
                    files.push_back({next_id++, path, size, sys_mtime, birth, ""});
                } else {
                    files.push_back({next_id++, path, size, sys_mtime, sys_mtime, ""});
                }
                scanned++;
            }
        } catch (const fs::filesystem_error& e) {
            std::cerr << "Warning: " << e.what() << std::endl;
        }
    }
    return files;
}

std::unordered_map<uint64_t, std::vector<FileInfo*>> group_by_size(std::vector<FileInfo>& files) {
    std::unordered_map<uint64_t, std::vector<FileInfo*>> groups;
    for (auto& f : files) {
        groups[f.size].push_back(&f);
    }
    return groups;
}

std::string csv_escape(const std::string& s) {
    if (s.find(',') != std::string::npos || s.find('"') != std::string::npos || s.find('\n') != std::string::npos) {
        std::string escaped = "\"";
        for (char c : s) {
            if (c == '"') escaped += "\"\"";
            else escaped += c;
        }
        escaped += "\"";
        return escaped;
    }
    return s;
}

std::string format_duration(double seconds) {
    int64_t ms_total = static_cast<int64_t>(seconds * 1000);
    int64_t ms = ms_total % 1000;
    int64_t s = ms_total / 1000;
    int64_t m = s / 60;
    s %= 60;
    int64_t h = m / 60;
    m %= 60;
    std::ostringstream oss;
    oss << std::setw(2) << std::setfill('0') << h << ":"
        << std::setw(2) << m << ":"
        << std::setw(2) << s << "."
        << std::setw(3) << ms;
    return oss.str();
}

std::string to_lower(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(), ::tolower);
    return s;
}

std::string get_category(const std::string& ext) {
    if (ext == ".c" || ext == ".h" || ext == ".cpp" || ext == ".hpp" || ext == ".cc" || ext == ".cxx" ||
        ext == ".m" || ext == ".mm" || ext == ".s" || ext == ".S" || ext == ".java" || ext == ".kt" ||
        ext == ".py" || ext == ".js" || ext == ".ts" || ext == ".rs" || ext == ".go" || ext == ".rb" ||
        ext == ".swift" || ext == ".sh") return "source";
    if (ext == ".hex" || ext == ".bin" || ext == ".elf" || ext == ".dfu" || ext == ".flash" || ext == ".map") return "firmware";
    if (ext == ".uvprojx" || ext == ".uvoptx" || ext == ".ewp" || ext == ".eww" || ext == ".ewt" ||
        ext == ".cproject" || ext == ".project" || ext == ".mxproject" || ext == ".ioc") return "ide";
    if (ext == ".yaml" || ext == ".yml" || ext == ".cmake" || ext == ".json" || ext == ".xml" ||
        ext == ".conf" || ext == ".cfg" || ext == ".ini" || ext == ".toml" || ext == ".properties") return "config";
    if (ext == ".pdf" || ext == ".md" || ext == ".txt" || ext == ".html" || ext == ".htm" ||
        ext == ".rst" || ext == ".chm" || ext == ".doc" || ext == ".docx" || ext == ".rtf") return "docs";
    if (ext == ".png" || ext == ".jpg" || ext == ".jpeg" || ext == ".gif" || ext == ".webp" ||
        ext == ".svg" || ext == ".bmp" || ext == ".ico" || ext == ".tiff" || ext == ".tif") return "image";
    if (ext == ".exe" || ext == ".dll" || ext == ".so" || ext == ".dylib" || ext == ".o" || ext == ".a" ||
        ext == ".lib" || ext == ".obj" || ext == ".gch" || ext == ".pch") return "binary";
    if (ext == ".zip" || ext == ".7z" || ext == ".tar" || ext == ".gz" || ext == ".bz2" ||
        ext == ".xz" || ext == ".rar" || ext == ".tgz") return "archive";
    if (ext == ".mp4" || ext == ".wav" || ext == ".avi" || ext == ".mp3" || ext == ".ogg" ||
        ext == ".flac" || ext == ".mov" || ext == ".wmv") return "media";
    if (ext == ".ttf" || ext == ".otf" || ext == ".woff" || ext == ".woff2") return "font";
    if (ext == ".csv" || ext == ".dts" || ext == ".dtsi" || ext == ".overlay" || ext == ".ld" ||
        ext == ".icf" || ext == ".srec") return "data";
    return "other";
}

void generate_analytics(const std::vector<FileInfo>& all_files,
                        const std::unordered_map<uint64_t, std::vector<FileInfo*>>& by_size,
                        double elapsed) {
    struct CatStats { uint64_t count = 0, total_bytes = 0, dup_count = 0, dup_bytes = 0; };
    struct ExtStats { uint64_t count = 0, total_bytes = 0, dup_count = 0, dup_bytes = 0; };

    std::unordered_map<std::string, CatStats> cat_map;
    std::unordered_map<std::string, ExtStats> ext_map;
    uint64_t total_size = 0, dup_files = 0, dup_bytes = 0;
    int size_bins[6] = {0, 0, 0, 0, 0, 0}; // 0b, <1k, 1k-100k, 100k-1m, 1m-100m, >100m

    for (const auto& f : all_files) {
        std::string ext_s = to_lower(f.path.extension().string());
        std::string cat = get_category(ext_s);
        cat_map[cat].count++;
        cat_map[cat].total_bytes += f.size;
        ext_map[ext_s].count++;
        ext_map[ext_s].total_bytes += f.size;
        total_size += f.size;
        if (f.size == 0) size_bins[0]++;
        else if (f.size < 1024) size_bins[1]++;
        else if (f.size < 102400) size_bins[2]++;
        else if (f.size < 1048576) size_bins[3]++;
        else if (f.size < 104857600) size_bins[4]++;
        else size_bins[5]++;
    }

    // Identify duplicates
    for (const auto& kv : by_size) {
        auto& group = kv.second;
        if (group.size() < 2) continue;
        std::unordered_map<Hash, std::vector<FileInfo*>> hg;
        for (auto* f : group) { if (!f->hash.empty()) hg[f->hash].push_back(f); }
        for (auto& h : hg) {
            if (h.second.size() < 2) continue;
            for (size_t i = 1; i < h.second.size(); ++i) {
                auto* f = h.second[i];
                std::string ext_s = f->path.extension().string();
                std::string cat = get_category(ext_s);
                cat_map[cat].dup_count++;
                cat_map[cat].dup_bytes += f->size;
                ext_map[ext_s].dup_count++;
                ext_map[ext_s].dup_bytes += f->size;
                dup_files++;
                dup_bytes += f->size;
            }
        }
    }

    uint64_t recoverable = dup_bytes;

    // Write JSON
    std::ofstream jf("analytics_cpp.json");
    jf << "{\n";
    jf << "  \"summary\": {";
    jf << "\"total_files\": " << all_files.size() << ", ";
    jf << "\"total_size_bytes\": " << total_size << ", ";
    jf << "\"duplicate_files\": " << dup_files << ", ";
    jf << "\"duplicate_size_bytes\": " << dup_bytes << ", ";
    jf << "\"recoverable_bytes\": " << recoverable << ", ";
    jf << "\"scan_duration_seconds\": " << std::fixed << std::setprecision(3) << elapsed << "},\n";

    jf << "  \"by_category\": {\n";
    bool first_cat = true;
    for (auto& kv : cat_map) {
        if (!first_cat) jf << ",\n";
        first_cat = false;
        jf << "    \"" << kv.first << "\": {";
        jf << "\"count\": " << kv.second.count << ", ";
        jf << "\"total_bytes\": " << kv.second.total_bytes << ", ";
        jf << "\"duplicate_count\": " << kv.second.dup_count << ", ";
        jf << "\"duplicate_bytes\": " << kv.second.dup_bytes << ", ";
        jf << "\"extensions\": {";
        bool first_ext = true;
        for (auto& ekv : ext_map) {
            if (get_category(ekv.first) != kv.first) continue;
            if (!first_ext) jf << ", ";
            first_ext = false;
            jf << "\"" << ekv.first << "\": " << ekv.second.count;
        }
        jf << "}}";
    }
    jf << "\n  },\n";

    jf << "  \"by_extension\": {\n";
    bool first_e = true;
    for (auto& kv : ext_map) {
        if (!first_e) jf << ",\n";
        first_e = false;
        jf << "    \"" << kv.first << "\": {";
        jf << "\"count\": " << kv.second.count << ", ";
        jf << "\"total_bytes\": " << kv.second.total_bytes << ", ";
        jf << "\"duplicate_count\": " << kv.second.dup_count << ", ";
        jf << "\"duplicate_bytes\": " << kv.second.dup_bytes << "}";
    }
    jf << "\n  },\n";

    jf << "  \"size_distribution\": {";
    jf << "\"0_bytes\": " << size_bins[0] << ", ";
    jf << "\"under_1kb\": " << size_bins[1] << ", ";
    jf << "\"1kb_100kb\": " << size_bins[2] << ", ";
    jf << "\"100kb_1mb\": " << size_bins[3] << ", ";
    jf << "\"1mb_100mb\": " << size_bins[4] << ", ";
    jf << "\"over_100mb\": " << size_bins[5] << "}\n";
    jf << "}\n";

    // Print summary
    std::cout << "\n--- File Type Analytics ---\n";
    for (auto& kv : cat_map) {
        std::cout << "  " << std::setw(10) << std::left << kv.first
                  << ": " << kv.second.count << " files, "
                  << kv.second.dup_count << " duplicates\n";
    }
    std::cout << "Analytics written to analytics_cpp.json\n";
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: find_dups_cpp <dir1> [<dir2> ...]" << std::endl;
        return 1;
    }

    std::vector<Path> roots;
    for (int i = 1; i < argc; ++i) roots.emplace_back(argv[i]);

    auto start_total = std::chrono::steady_clock::now();

    std::cout << "Collecting files... " << std::flush;
    uint64_t scanned = 0;
    auto all_files = collect_files(roots, scanned);
    std::cout << "found " << scanned << " files" << std::endl;
    if (scanned == 0) return 0;

    auto by_size = group_by_size(all_files);

    // --- Collect files to hash (size groups with >1 file) ---
    std::vector<FileInfo*> files_to_hash;
    for (auto& kv : by_size) {
        if (kv.second.size() > 1) {
            files_to_hash.insert(files_to_hash.end(), kv.second.begin(), kv.second.end());
        }
    }

    if (!files_to_hash.empty()) {
        std::cout << "Parallel hashing " << files_to_hash.size()
                  << " files (workers: " << NUM_WORKERS << ")..." << std::endl;
        auto start_hash = std::chrono::steady_clock::now();

        auto hash_worker = [](std::vector<FileInfo*>* chunk) {
            for (auto* f : *chunk) {
                f->hash = compute_sha256(f->path);
            }
        };

        size_t chunk_size = (files_to_hash.size() + NUM_WORKERS - 1) / NUM_WORKERS;
        std::vector<std::vector<FileInfo*>> chunks;
        for (size_t i = 0; i < files_to_hash.size(); i += chunk_size) {
            chunks.emplace_back(files_to_hash.begin() + i,
                                files_to_hash.begin() + std::min(i + chunk_size, files_to_hash.size()));
        }
        std::vector<std::future<void>> futures;
        for (auto& chunk : chunks) {
            futures.push_back(std::async(std::launch::async, hash_worker, &chunk));
        }
        for (auto& f : futures) f.get();

        auto end_hash = std::chrono::steady_clock::now();
        std::chrono::duration<double> hash_elapsed = end_hash - start_hash;
        std::cout << "Hashing completed in " << hash_elapsed.count() << " seconds" << std::endl;
    }

    // ----- Duplicate detection -----
    std::vector<std::vector<std::string>> csv_dups;
    std::vector<std::string> bash_lines = {"#!/bin/bash", "# Generated by find_dups_cpp", "set -e", ""};
    uint64_t duplicates_count = 0;

    for (auto& kv : by_size) {
        auto& group = kv.second;
        if (group.size() < 2) continue;

        std::unordered_map<Hash, std::vector<FileInfo*>> hash_groups;
        for (auto* f : group) {
            if (!f->hash.empty()) hash_groups[f->hash].push_back(f);
        }
        for (auto& hg : hash_groups) {
            auto& same = hg.second;
            if (same.size() < 2) continue;
            std::sort(same.begin(), same.end(), [](auto a, auto b) { return a->id < b->id; });
            for (auto* f : same) {
                csv_dups.push_back({
                    std::to_string(f->id),
                    f->path.string(),
                    std::to_string(f->size),
                    f->hash,
                    format_time(f->birth_time),
                    format_time(f->mod_time)
                });
            }
            for (size_t i = 1; i < same.size(); ++i) {
                duplicates_count++;
                std::string escaped = same[i]->path.string();
                size_t pos = 0;
                while ((pos = escaped.find('\'', pos)) != std::string::npos) {
                    escaped.replace(pos, 1, "'\\''");
                    pos += 4;
                }
                bash_lines.push_back("rm -- '" + escaped + "'");
            }
        }
    }
    bash_lines.push_back("");
    bash_lines.push_back("echo \"Deletion complete.\"");

    std::ofstream csv_file("duplicates_cpp.csv");
    if (csv_file) {
        csv_file << "FileID,Path,Size,Hash,CreationTime,ModificationTime\n";
        for (auto& row : csv_dups) {
            for (size_t i = 0; i < row.size(); ++i) {
                if (i) csv_file << ',';
                csv_file << csv_escape(row[i]);
            }
            csv_file << '\n';
        }
    } else {
        std::cerr << "Error creating duplicates_cpp.csv\n";
        return 1;
    }

    std::ofstream sh_file("duprm_cpp.sh");
    if (sh_file) {
        for (auto& line : bash_lines) sh_file << line << '\n';
        sh_file.close();
        fs::permissions("duprm_cpp.sh", fs::perms::owner_exec | fs::perms::group_exec | fs::perms::others_exec,
                        fs::perm_options::add);
    } else {
        std::cerr << "Error creating duprm_cpp.sh\n";
    }

    std::vector<FileInfo> sorted = all_files;
    std::sort(sorted.begin(), sorted.end(), [](auto& a, auto& b) { return a.size > b.size; });
    std::ofstream sort_file("sort_dup_cpp.csv");
    if (sort_file) {
        sort_file << "FileID,Path,Size,Hash,CreationTime,ModificationTime\n";
        for (auto& f : sorted) {
            sort_file << csv_escape(std::to_string(f.id)) << ','
                      << csv_escape(f.path.string()) << ','
                      << csv_escape(std::to_string(f.size)) << ','
                      << csv_escape(f.hash) << ','
                      << csv_escape(format_time(f.birth_time)) << ','
                      << csv_escape(format_time(f.mod_time)) << '\n';
        }
    } else {
        std::cerr << "Error creating sort_dup_cpp.csv\n";
    }

    auto end_total = std::chrono::steady_clock::now();
    Duration elapsed = end_total - start_total;
    double sec = elapsed.count();

    generate_analytics(all_files, by_size, sec);

    std::cout << "\nFiles scanned: " << scanned << std::endl;
    std::cout << "Duplicates found (files to delete): " << duplicates_count << std::endl;
    std::cout << "Runtime:\n";
    std::cout << "  - " << std::fixed << std::setprecision(3) << sec << " seconds\n";
    std::cout << "  - " << format_duration(sec) << std::endl;
    std::cout << "Reports: duplicates_cpp.csv, sort_dup_cpp.csv\n";
    std::cout << "Delete script: duprm_cpp.sh\n";

    return 0;
}

File: find_dups_pthon/find_dups.py
#!/usr/bin/env python3
"""find_dups.py - Duplicate file finder (regular files only, no symlinks)."""

import os
import sys
import hashlib
import csv
import json
import time
from multiprocessing import Pool, cpu_count
from collections import defaultdict

BUFFER_SIZE = 64 * 1024
NUM_WORKERS = cpu_count()

EXTENSION_CATEGORIES = {
    '.c': 'source', '.h': 'source', '.cpp': 'source', '.hpp': 'source',
    '.cc': 'source', '.cxx': 'source', '.m': 'source', '.mm': 'source',
    '.s': 'source', '.S': 'source', '.java': 'source', '.kt': 'source',
    '.py': 'source', '.js': 'source', '.ts': 'source', '.rs': 'source',
    '.go': 'source', '.rb': 'source', '.swift': 'source', '.sh': 'source',
    '.hex': 'firmware', '.bin': 'firmware', '.elf': 'firmware', '.dfu': 'firmware',
    '.flash': 'firmware', '.map': 'firmware',
    '.uvprojx': 'ide', '.uvoptx': 'ide', '.ewp': 'ide', '.eww': 'ide',
    '.ewt': 'ide', '.cproject': 'ide', '.project': 'ide', '.mxproject': 'ide',
    '.ioc': 'ide',
    '.yaml': 'config', '.yml': 'config', '.cmake': 'config', '.json': 'config',
    '.xml': 'config', '.conf': 'config', '.cfg': 'config', '.ini': 'config',
    '.toml': 'config', '.properties': 'config',
    '.pdf': 'docs', '.md': 'docs', '.txt': 'docs', '.html': 'docs',
    '.htm': 'docs', '.rst': 'docs', '.chm': 'docs', '.doc': 'docs',
    '.docx': 'docs', '.rtf': 'docs',
    '.png': 'image', '.jpg': 'image', '.jpeg': 'image', '.gif': 'image',
    '.webp': 'image', '.svg': 'image', '.bmp': 'image', '.ico': 'image',
    '.tiff': 'image', '.tif': 'image',
    '.exe': 'binary', '.dll': 'binary', '.so': 'binary', '.dylib': 'binary',
    '.o': 'binary', '.a': 'binary', '.lib': 'binary', '.obj': 'binary',
    '.gch': 'binary', '.pch': 'binary',
    '.zip': 'archive', '.7z': 'archive', '.tar': 'archive', '.gz': 'archive',
    '.bz2': 'archive', '.xz': 'archive', '.rar': 'archive', '.tgz': 'archive',
    '.mp4': 'media', '.wav': 'media', '.avi': 'media', '.mp3': 'media',
    '.ogg': 'media', '.flac': 'media', '.mov': 'media', '.wmv': 'media',
    '.ttf': 'font', '.otf': 'font', '.woff': 'font', '.woff2': 'font',
    '.csv': 'data', '.dts': 'data', '.dtsi': 'data', '.overlay': 'data',
    '.ld': 'data', '.icf': 'data', '.srec': 'data',
}

def is_regular_file(path):
    try:
        st = os.lstat(path)
        return (st.st_mode & 0o170000) == 0o100000  # S_IFREG
    except OSError:
        return False

def collect_files(roots):
    files = []
    next_id = 1
    for root in roots:
        root_path = os.path.abspath(os.path.expanduser(root))
        if not os.path.exists(root_path):
            print(f"Warning: {root_path} does not exist", file=sys.stderr)
            continue
        for dirpath, dirnames, filenames in os.walk(root_path, followlinks=False):
            for fn in filenames:
                full = os.path.join(dirpath, fn)
                if is_regular_file(full):
                    try:
                        st = os.lstat(full)
                    except OSError:
                        continue
                    if st.st_size == 0:
                        continue  # skip zero-byte files
                    files.append({
                        'id': next_id,
                        'path': full,
                        'size': st.st_size,
                        'mtime': st.st_mtime,
                        'ctime': st.st_birthtime if hasattr(st, 'st_birthtime') else st.st_ctime,
                        'hash': None
                    })
                    next_id += 1
    return files

def compute_sha256(args):
    file_path, _ = args
    try:
        sha = hashlib.sha256()
        with open(file_path, 'rb') as f:
            for chunk in iter(lambda: f.read(BUFFER_SIZE), b''):
                sha.update(chunk)
        return sha.hexdigest()
    except Exception:
        return None

def parallel_hash_chunk(chunk):
    results = []
    for f in chunk:
        f['hash'] = compute_sha256((f['path'], None))
        results.append((f['id'], f['hash']))
    return results

def format_duration(seconds):
    ms = int(seconds * 1000)
    s, ms = divmod(ms, 1000)
    m, s = divmod(s, 60)
    h, m = divmod(m, 60)
    return f"{h:02d}:{m:02d}:{s:02d}.{ms:03d}"

def write_csv(filepath, rows, header):
    try:
        with open(filepath, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(header)
            writer.writerows(rows)
        return True
    except Exception as e:
        print(f"Error writing {filepath}: {e}", file=sys.stderr)
        return False

def generate_analytics(all_files, dup_groups, elapsed):
    """Generate analytics JSON and print summary."""
    # Build set of duplicate file ids (files to delete = all but first in each group)
    dup_ids = set()
    for _size, group, _h in dup_groups:
        for f in group[1:]:
            dup_ids.add(f['id'])

    total_size = sum(f['size'] for f in all_files)
    dup_size = sum(f['size'] for f in all_files if f['id'] in dup_ids)

    # Helper to get extension and category
    def get_ext(file_path):
        basename = os.path.basename(file_path)
        dot = basename.rfind('.')
        if dot > 0:
            return basename[dot:].lower()
        elif dot == 0 and len(basename) > 1:
            return basename.lower()
        return '(none)'

    # by_extension
    by_ext = defaultdict(lambda: {'count': 0, 'total_bytes': 0, 'duplicate_count': 0, 'duplicate_bytes': 0})
    for f in all_files:
        ext = get_ext(f['path'])
        e = by_ext[ext]
        e['count'] += 1
        e['total_bytes'] += f['size']
        if f['id'] in dup_ids:
            e['duplicate_count'] += 1
            e['duplicate_bytes'] += f['size']

    # by_category
    by_cat = defaultdict(lambda: {'count': 0, 'total_bytes': 0, 'duplicate_count': 0, 'duplicate_bytes': 0, 'extensions': {}})
    for ext, edata in by_ext.items():
        cat = EXTENSION_CATEGORIES.get(ext, 'other')
        c = by_cat[cat]
        c['count'] += edata['count']
        c['total_bytes'] += edata['total_bytes']
        c['duplicate_count'] += edata['duplicate_count']
        c['duplicate_bytes'] += edata['duplicate_bytes']
        c['extensions'][ext] = edata

    # size_distribution
    size_bins = {'0_bytes': 0, 'under_1kb': 0, '1kb_100kb': 0, '100kb_1mb': 0, '1mb_100mb': 0, 'over_100mb': 0}
    for f in all_files:
        s = f['size']
        if s == 0:
            size_bins['0_bytes'] += 1
        elif s < 1024:
            size_bins['under_1kb'] += 1
        elif s < 100 * 1024:
            size_bins['1kb_100kb'] += 1
        elif s < 1024 * 1024:
            size_bins['100kb_1mb'] += 1
        elif s < 100 * 1024 * 1024:
            size_bins['1mb_100mb'] += 1
        else:
            size_bins['over_100mb'] += 1

    analytics = {
        'summary': {
            'total_files': len(all_files),
            'total_size_bytes': total_size,
            'duplicate_files': len(dup_ids),
            'duplicate_size_bytes': dup_size,
            'recoverable_bytes': dup_size,
            'scan_duration_seconds': round(elapsed, 3),
        },
        'by_category': dict(by_cat),
        'by_extension': dict(by_ext),
        'size_distribution': size_bins,
    }

    with open('analytics_py.json', 'w') as jf:
        json.dump(analytics, jf, indent=2)

    # Print human-readable summary
    print("\n--- Analytics Summary ---")
    print(f"Total files: {len(all_files)}  |  Total size: {total_size:,} bytes")
    print(f"Duplicate files (to delete): {len(dup_ids)}  |  Recoverable: {dup_size:,} bytes")
    print(f"Scan duration: {elapsed:.3f}s")
    print("\nBy category:")
    for cat in sorted(by_cat):
        c = by_cat[cat]
        print(f"  {cat:12s}  {c['count']:6d} files  {c['total_bytes']:>14,} bytes  "
              f"({c['duplicate_count']} dups, {c['duplicate_bytes']:,} bytes)")
    print("\nBy extension (top 15 by count):")
    sorted_exts = sorted(by_ext.items(), key=lambda x: x[1]['count'], reverse=True)[:15]
    for ext, edata in sorted_exts:
        print(f"  {ext:12s}  {edata['count']:6d} files  {edata['total_bytes']:>14,} bytes  "
              f"({edata['duplicate_count']} dups)")
    print(f"\nSize distribution: {dict(size_bins)}")
    print("Analytics saved: analytics_py.json")

def main():
    if len(sys.argv) < 2:
        print("Usage: python3 find_dups.py <dir1> [<dir2> ...]")
        sys.exit(1)

    start_total = time.time()
    roots = sys.argv[1:]

    print("Collecting files...", end=' ', flush=True)
    all_files = collect_files(roots)
    total_scanned = len(all_files)
    print(f"found {total_scanned} files")
    if total_scanned == 0:
        return

    # Group by size
    by_size = defaultdict(list)
    for f in all_files:
        by_size[f['size']].append(f)

    # Files to hash (size not unique)
    files_to_hash = []
    for group in by_size.values():
        if len(group) > 1:
            files_to_hash.extend(group)
    total_to_hash = len(files_to_hash)

    if total_to_hash > 0:
        print(f"Parallel hashing {total_to_hash} files (workers: {NUM_WORKERS})...")
        start_hash = time.time()

        chunk_size = max(1, total_to_hash // (NUM_WORKERS * 4))
        chunks = [files_to_hash[i:i+chunk_size] for i in range(0, total_to_hash, chunk_size)]

        with Pool(NUM_WORKERS) as pool:
            results = pool.map(parallel_hash_chunk, chunks)

        # Apply hashes from workers
        hash_map = {}
        for chunk_result in results:
            for fid, fhash in chunk_result:
                hash_map[fid] = fhash
        for f in files_to_hash:
            if f['id'] in hash_map:
                f['hash'] = hash_map[f['id']]

        print(f"Hashing completed in {format_duration(time.time() - start_hash)}")

    # Build duplicate groups
    dup_groups = []
    for size, group in by_size.items():
        if len(group) < 2:
            continue
        hash_groups = defaultdict(list)
        for f in group:
            if f['hash']:
                hash_groups[f['hash']].append(f)
        for h, g in hash_groups.items():
            if len(g) > 1:
                g.sort(key=lambda x: x['id'])
                dup_groups.append((size, g, h))

    # CSV dups
    dup_rows = []
    bash_lines = ["#!/bin/bash", "# Generated by find_dups.py", "set -e", ""]
    dups_to_delete = 0
    for size, group, h in dup_groups:
        for f in group:
            ctime = time.strftime('%Y-%m-%dT%H:%M:%S', time.localtime(f['ctime'])) + 'Z'
            mtime = time.strftime('%Y-%m-%dT%H:%M:%S', time.localtime(f['mtime'])) + 'Z'
            dup_rows.append([f['id'], f['path'], size, h, ctime, mtime])
        for f in group[1:]:
            dups_to_delete += 1
            escaped = f['path'].replace("'", "'\\''")
            bash_lines.append(f"rm -- '{escaped}'")
    bash_lines.append('')
    bash_lines.append('echo "Deletion complete."')

    write_csv('duplicates_py.csv', dup_rows,
              ['FileID', 'Path', 'Size', 'Hash', 'CreationTime', 'ModificationTime'])
    with open('duprm_py.sh', 'w') as f:
        f.write('\n'.join(bash_lines))
    os.chmod('duprm_py.sh', 0o755)

    # Sort CSV
    sorted_files = sorted(all_files, key=lambda x: x['size'], reverse=True)
    sort_rows = []
    for f in sorted_files:
        ctime = time.strftime('%Y-%m-%dT%H:%M:%S', time.localtime(f['ctime'])) + 'Z'
        mtime = time.strftime('%Y-%m-%dT%H:%M:%S', time.localtime(f['mtime'])) + 'Z'
        sort_rows.append([f['id'], f['path'], f['size'], f['hash'] or '', ctime, mtime])
    write_csv('sort_dup_py.csv', sort_rows,
              ['FileID', 'Path', 'Size', 'Hash', 'CreationTime', 'ModificationTime'])

    elapsed = time.time() - start_total
    generate_analytics(all_files, dup_groups, elapsed)
    print(f"\nFiles scanned: {total_scanned}")
    print(f"Duplicates found (files to delete): {dups_to_delete}")
    print(f"Runtime:")
    print(f"  - {elapsed:.3f} seconds")
    print(f"  - {format_duration(elapsed)}")
    print("Reports: duplicates_py.csv, sort_dup_py.csv")
    print("Delete script: duprm_py.sh")

if __name__ == '__main__':
    main()
